src/Resources/contao/library/Contao/System.php

```diff
@@ -562,32 +562,74 @@ public static function getTimeZones()
 	/**
 	 * Return all image sizes as array
 	 *
+	 * @param bool $all Returns all sizes omitting the back end user permission check
+	 *
 	 * @return array The available image sizes
 	 */
-	public static function getImageSizes()
+	public static function getImageSizes($all = false)
 	{
-		if (empty(static::$arrImageSizes))
-		{
-			try
-			{
+		if (empty(static::$arrImageSizes)) {
+			try {
 				$sizes = array();
-				$imageSize = \Database::getInstance()->query("SELECT id, name, width, height FROM tl_image_size ORDER BY pid, name");
+				$imageSize = \Database::getInstance()->query(
+					"SELECT id, name, width, height FROM tl_image_size ORDER BY pid, name"
+				);
 
-				while ($imageSize->next())
-				{
+				while ($imageSize->next()) {
 					$sizes[$imageSize->id] = $imageSize->name;
 					$sizes[$imageSize->id] .= ' (' . $imageSize->width . 'x' . $imageSize->height . ')';
 				}
 
 				static::$arrImageSizes = array_merge(array('image_sizes' => $sizes), $GLOBALS['TL_CROP']);
-			}
-			catch (\Exception $e)
-			{
+			} catch (\Exception $e) {
 				static::$arrImageSizes = $GLOBALS['TL_CROP'];
 			}
 		}
 
-		return static::$arrImageSizes;
+		$imageSizes = static::$arrImageSizes;
+
+		// Permission check
+		if (TL_MODE == 'BE' && $all === false) {
+			$user = \BackendUser::getInstance();
+
+			// Limit only if the user is not an admin
+			if (!$user->isAdmin) {
+				$allowedSizes = deserialize($user->imageSizes, true);
+
+				if (!empty($allowedSizes)) {
+					$filteredSizes = [];
+
+					foreach ($imageSizes as $group => $sizes) {
+						foreach ($sizes as $id => $size) {
+
+							// Dynamic sizes
+							if ($group == 'image_sizes') {
+								if (in_array($id, $allowedSizes)) {
+									$filteredSizes[$group][$id] = $size;
+								}
+
+								continue;
+							}
+
+							if (in_array($size, $allowedSizes)) {
+								$filteredSizes[$group][] = $size;
+							}
+						}
+					}
+
+					$imageSizes = $filteredSizes;
+				}
+			}
+		}
+
+		// HOOK: allow to add custom logic
+		if (isset($GLOBALS['TL_HOOKS']['getImageSizes']) && is_array($GLOBALS['TL_HOOKS']['getImageSizes'])) {
+			foreach ($GLOBALS['TL_HOOKS']['getImageSizes'] as $callback) {
+				$imageSizes = static::importStatic($callback[0])->$callback[1]($imageSizes);
```

Comment:
It is, for converting the existing hooks. But I would not add new ones. I'm open to other opinions though :-)